## Three-Point Endpoint Formula & Richardson Extrapolation


**Name: Jiaen Qi**

**Date: 08/23/2026**




## Setup

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Three-Point Endpoint Formula

## Explain

The three-point endpoint formula is

$$
D_h f(x_0)
=
\frac{-3f(x_0)+4f(x_0+h)-f(x_0+2h)}{2h}.
$$

Using Taylor expansions about $$x_0$$,

$$
f(x_0+h)
=
f(x_0)+hf'(x_0)
+\frac{h^2}{2}f''(x_0)
+\frac{h^3}{6}f^{(3)}(x_0)
+O(h^4),
$$

and

$$
f(x_0+2h)
=
f(x_0)+2hf'(x_0)
+2h^2f''(x_0)
+\frac{4}{3}h^3f^{(3)}(x_0)
+O(h^4).
$$

Plugging these into the formula, we have

$$
D_h f(x_0)
=
f'(x_0)
-
\frac{h^2}{3}f^{(3)}(x_0)
+
O(h^3).
$$

Therefore,

$$
f'(x_0)-D_hf(x_0)
=
\frac{h^2}{3}f^{(3)}(\xi),
\qquad
\xi\in(x_0,x_0+2h).
$$

Since the error contains a factor of $h^2$, the formula has error

$$
O(h^2)
$$
.

##  Implement

We will define the function by the following equation, which is the three-point endpoint formula

$$
D_h f(x_0)
=
\frac{-3f(x_0)+4f(x_0+h)-f(x_0+2h)}{2h}.
$$

It has the point $x_0$, and the step size $h$ as inputs.

In [ ]:
def three_point_endpoint(f, x0, h):
    # Evaluate f at the three required points
    fx0 = f(x0)
    fx1 = f(x0 + h)
    fx2 = f(x0 + 2*h)

    # Apply the three-point endpoint formula
    derivative = (-3*fx0 + 4*fx1 - fx2) / (2*h)

    return derivative

## Experiment and Example

For

$$
f(x)=xe^x,
$$

the exact derivative is

$$
f'(x)=(x+1)e^x.
$$

At $$x_0=1$$,

$$
f'(1)=2e.
$$

To find the error bound, we need the third derivative:

$$
f''(x)=(x+2)e^x,
$$

and

$$
f^{(3)}(x)=(x+3)e^x.
$$

The error formula is

$$
\left|f'(x_0)-D_hf(x_0)\right|
\leq
\frac{h^2}{3}
\max_{\xi\in[x_0,x_0+2h]}
\left|f^{(3)}(\xi)\right|.
$$

Since

$$
f^{(3)}(x)=(x+3)e^x
$$

is increasing for the intervals used here, its maximum occurs at

$$
x=x_0+2h.
$$

Therefore, the error bound is

$$
\frac{h^2}{3}(x_0+2h+3)e^{x_0+2h}.
$$

In [ ]:
def f(x):
  return x * np.exp(x)

# First derivative
def f_prime(x):
  return (x + 1) * np.exp(x)

# Third derivative
def f_third(x):
  return (x + 3) * np.exp(x)

# Experiment values
x0 = 1.0
h_values = [0.1, 0.05]

results = []

for h in h_values:
  fx0 = f(x0)
  fx1 = f(x0 + h)
  fx2 = f(x0 + 2*h)

  approximation = three_point_endpoint(f, x0, h)

  exact_value = f_prime(x0)
  actual_error = abs(exact_value - approximation)

  # f''' is increasing, so its maximum occurs at x0 + 2h
  error_bound = (h**2 / 3) * f_third(x0 + 2*h)



# Then we make a table

  results.append({
      "h":h,
      "f(x0)": fx0,
      "f(x0+h)": fx1,
      "f(x0+2h)": fx2,
      "D_h f(x0)": approximation,
      "f'(x0)": exact_value,
      "actual error": actual_error,
      "error bound": error_bound
  })

results_table = pd.DataFrame(results)
results_table

,h,f(x0),f(x0+h),f(x0+2h),D_h f(x0),f'(x0),actual error,error bound
0,0.10,2.718282,3.304583,3.984140,5.396724,5.436564,0.039840,0.046482
1,0.05,2.718282,3.000534,3.304583,5.427066,5.436564,0.009498,0.010264


When h decreases from 0.1 to 0.05, the actual error decreases from approximately 0.039840 to 0.009498. Thus,

$$
\frac{0.039840}{0.009498} \approx 4.19.
$$

Therefore, h reduces the error by approximately a factor of 4. This agrees with the expected $O(h^2)$ accuracy because

$$
\left(\frac{h}{2}\right)^2=\frac{h^2}{4}.
$$

Also, the actual error is smaller than the calculated error bound for both step sizes.

## Conclusion

The three-point endpoint formula produced approximations close to the exact value $f'(1)=2e$.  
When $h$ was halved from $0.1$ to $0.05$, the actual error decreased by a factor of approximately $4.19$.  
Both actual errors were below their corresponding theoretical error bounds.  
Therefore, the experiment agrees with the expected second-order, $O(h^2)$, accuracy.

## Richardson Extrapolation

## Explain

The centered-difference formula is

$$
D(h)=\frac{f(x+h)-f(x-h)}{2h}.
$$

Taylor expansion shows that

$$
D(h)=f'(x)+Ch^2+O(h^4),
$$

where $Ch^2$ is the leading error term. If the step size is halved, then

$$
D(h/2)=f'(x)+\frac{Ch^2}{4}+O(h^4).
$$

Multiplying this equation by $4$ gives

$$
4D(h/2)=4f'(x)+Ch^2+O(h^4).
$$

Subtracting $D(h)$ cancels the matching $Ch^2$ terms:

$$
4D(h/2)-D(h)=3f'(x)+O(h^4).
$$

Therefore,

$$
R(h)=\frac{4D(h/2)-D(h)}{3}
     =f'(x)+O(h^4).
$$

`Essentially`, Richardson extrapolation cancels the leading $O(h^2)$ error and improves the centered-difference formula to $O(h^4)$ accuracy.

## Implement

Then we will define the function in python with the following codes:

In [ ]:
def richardson_extrapolation(f, x, h):
    # Centered-difference approximation using step size h
    D_h = (f(x + h) - f(x - h)) / (2*h)

    # Centered-difference approximation using the smaller step size h/2
    half_h = h / 2
    D_half_h = (f(x + half_h) - f(x - half_h)) / (2*half_h)

    # Combine the approximations to cancel the leading O(h^2) error
    R_h = (4*D_half_h - D_h) / 3

    # Return all three values so their accuracy can be compared
    return D_h, D_half_h, R_h

## Experiment and Example

We use $h=0.1$ to approximate the derivatives at $x=1$ and $x=1.2$. For each point, we calculate the centered differences $D(h)$ and $D(h/2)$, followed by the Richardson approximation

$$
R(h)=\frac{4D(h/2)-D(h)}{3}.
$$

Each approximation is compared with the exact derivative

$$
f'(x)=(x+1)e^x,
$$

and its absolute error is calculated using

$$
\text{absolute error}
=
\left|\text{exact value}-\text{approximation}\right|.
$$

In [ ]:
h = 0.1
x_values = [1.0, 1.2]
richardson_results = []

for x in x_values:
    # Compute the three numerical approximations
    D_h, D_half_h, R_h = richardson_extrapolation(f, x, h)

    # Compute the exact derivative
    exact_value = f_prime(x)

    # Store the approximations and their absolute errors
    richardson_results.append({
        "x": x,
        "D(h)": D_h,
        "error D(h)": abs(exact_value - D_h),
        "D(h/2)": D_half_h,
        "error D(h/2)": abs(exact_value - D_half_h),
        "R(h)": R_h,
        "error R(h)": abs(exact_value - R_h),
        "exact derivative": exact_value
    })

# Create and display the table
richardson_table = pd.DataFrame(richardson_results)
richardson_table

,x,D(h),error D(h),D(h/2),error D(h/2),R(h),error R(h),exact derivative
0,1.0,5.454699,0.018135,5.441095,0.004531,5.436560,0.000003,5.436564
1,1.2,7.327515,0.023258,7.310069,0.005811,7.304253,0.000004,7.304257


`Note`: At both $x=1$ and $x=1.2$, reducing the step size from $h$ to $h/2$ makes the centered-difference error approximately four times smaller. This agrees with its $O(h^2)$ accuracy.

The error in $R(h)$ is much smaller than the errors in both $D(h)$ and $D(h/2)$. Therefore, Richardson extrapolation significantly improved the approximation by canceling the leading $O(h^2)$ error term and producing $O(h^4)$ accuracy.

## Conclusion

Now let's see the comparison between the two methods.

The three-point endpoint formula is easier to implement and uses function values on only one side of $x$.  

However, its accuracy is limited to $O(h^2)$.  
Richardson extrapolation requires two centered-difference calculations and values on both sides of $x$.  

In return, it cancels the leading error and achieves the more accurate order $O(h^4)$.